lib/applications/resolvers.ts

```diff
@@ -713,3 +717,131 @@ export const generateApplicationsReport: Resolver = async (_, args, { prisma })
     ok: true,
   };
 };
+
+export const generatePermitHoldersReport: Resolver = async (_, args, { prisma }) => {
+  const {
+    input: { startDate, endDate, columns },
+  } = args;
+
+  const columnsSet = new Set(columns);
+
+  const applications = await prisma.application.findMany({
+    where: {
+      createdAt: {
+        gte: startDate,
+        lte: endDate,
+      },
+    },
+    select: {
+      rcdUserId: columnsSet.has(PermitHoldersReportColumn.UserId),
+      firstName: columnsSet.has(PermitHoldersReportColumn.ApplicantName),
+      middleName: columnsSet.has(PermitHoldersReportColumn.ApplicantName),
+      lastName: columnsSet.has(PermitHoldersReportColumn.ApplicantName),
+      dateOfBirth: columnsSet.has(PermitHoldersReportColumn.ApplicantDateOfBirth),
+      addressLine1: columnsSet.has(PermitHoldersReportColumn.HomeAddress),
+      addressLine2: columnsSet.has(PermitHoldersReportColumn.HomeAddress),
+      city: columnsSet.has(PermitHoldersReportColumn.HomeAddress),
+      province: columnsSet.has(PermitHoldersReportColumn.HomeAddress),
+      postalCode: columnsSet.has(PermitHoldersReportColumn.HomeAddress),
+      email: columnsSet.has(PermitHoldersReportColumn.Email),
+      phone: columnsSet.has(PermitHoldersReportColumn.PhoneNumber),
+      guardianFirstName: columnsSet.has(PermitHoldersReportColumn.GuardianPoaName),
+      guardianMiddleName: columnsSet.has(PermitHoldersReportColumn.GuardianPoaName),
+      guardianLastName: columnsSet.has(PermitHoldersReportColumn.GuardianPoaName),
+      guardianRelationship: columnsSet.has(PermitHoldersReportColumn.GuardianPoaRelation),
+      guardianAddressLine1: columnsSet.has(PermitHoldersReportColumn.GuardianPoaAddress),
+      guardianAddressLine2: columnsSet.has(PermitHoldersReportColumn.GuardianPoaAddress),
+      guardianPostalCode: columnsSet.has(PermitHoldersReportColumn.GuardianPoaAddress),
+      guardianCity: columnsSet.has(PermitHoldersReportColumn.GuardianPoaAddress),
+      guardianProvince: columnsSet.has(PermitHoldersReportColumn.GuardianPoaAddress),
+      permit: {
+        select: {
+          rcdPermitId: true,
+          active: columnsSet.has(PermitHoldersReportColumn.UserStatus),
+        },
+      },
+      permitType: columnsSet.has(PermitHoldersReportColumn.RecentAppType),
+    },
+  });
+
+  // Formats the date fields and adds totalAmount, applicantName and rcdPermitId properties to allow for csv writing
+  const csvApplications = applications.map(application => {
```

Comment:
Addressed!

> We should either 1) modify the former step to just select all columns (so that latter step can reliably format all the required fields), or 2) build the objects using a bunch of if conditions so account for some fields not being selected in the former step. While approach 1 gives better performance, approach 2 is easier to understand and maintain. For now, to prioritize simplicity, let's just select all the columns and all necessary nested columns, and keep the current logic for the mapping.


Just to clarify,  is this supposed to be the opposite (approach 2 gives better performance, and approach 1 is easier to understand)?

Reply:
Hmm thought about this a bit more there doesn't seem to be an issue with the original implementation, calling the mapping function on fields that we never selected doesn't break anything and returns mostly null values (which won't be written to the csv since they were never selected). 

So with the original implementation we have both better performance (since we only select/fetch the fields we want) and also have simplicity since we don't write if statements

Reply:
> Just to clarify, is this supposed to be the opposite (approach 2 gives better performance, and approach 1 is easier to understand)?
Sorry for the late response, you're correct in this statement! Sorry for the confusion.

My original issue was that we were using `columnsSet.has` for every `select` which is very verbose. But I think your current implementation has a good balance between readability and performance, so let's stick with that. Thanks for pointing this out :)